### PR TITLE
Http client agnosticism

### PR DESCRIPTION
**Pull request by tjarvstrand**
Allow Katt to function when embedded into code bases that use other http clients than hackney (such as ones that are still stuck with lhttpc :P)

### CODE REVIEW

File: src/katt.erl [hunk 1]
```diff
@@ -78,13 +84,15 @@ run(Scenario, ScenarioParams, ScenarioCallbacks) ->
   ProgressFun = proplists:get_value( progress
                                    , Callbacks
                                    ),
-  spawn_link(?MODULE, run, [self(), Scenario, Params, Callbacks]),
+  HttpClient = proplists:get_value(http_client, Options),
+  ok = HttpClient:init([]),
```
Comment (gadgetci) [inline src/katt.erl]: According to **Elvis**:
> Remove the dynamic function call on line 88. Only modules that define callbacks should make dynamic calls.

File: src/katt_callbacks.erl [hunk 2]
```diff
@@ -250,7 +252,8 @@ http_request( #katt_request{ method = Method
   Hdrs1 = proplists:delete("x-katt-sleep", Hdrs0),
   Hdrs = proplists:delete("x-katt-timeout", Hdrs1),
   timer:sleep(Sleep),
-  katt_util:external_http_request(Url, Method, Hdrs, Body, Timeout, []).
+  Client = proplists:get_value(http_client, Options),
+  Client:request(Method, Url, Hdrs, Body, Timeout).
```
Comment (gadgetci) [inline src/katt_callbacks.erl]: According to **Elvis**:
> Remove the dynamic function call on line 256. Only modules that define callbacks should make dynamic calls.

**Comment by andreineculau** [comment]
:) ... I'll have a look in the coming days

**Comment by tjarvstrand** [comment]
Thanks! :)

**Comment by andreineculau** [comment]
Out of curiousity though, was there something wrong/different with hackney that doesn't fit your codebase? Or you just don't want to bring in another app?

**Comment by andreineculau** [comment]
ok, so I've looked and have some questions:

1. if you want to keep to lhttpc, then why not just revert d9f4ede (trivial conflicts) in your codebase and keep it along while still merging in upstream changes (the probability of conflicts is small)?
2. why implement the `http_client` *option*, rather than pass on a custom `request` *callback*?

**Comment by tjarvstrand** [comment]
No, there's nothing wrong with Hackney that I'm aware of. The reason is that we already have a quite bloated dependency tree and we're trying to not make it explode too much for many reasons. Just to name few: Avoiding sub-dependency conflicts, code footprint (build time etc.), not forcing devs to familiarize themselves with many libraries that do the same thing :)

In general I think it's good to stick to strict vanilla upstream whenever possible to minimize the mental overhead of keeping our dependencies up to date. For a single library it may not be too much, but it adds up.

I chose to go with a module based callback because the client may need some initialization to take place and I'd rather not do that at every call. We're using katt for performance tests at Klarna so I'd like to keep the overhead to a minimum (I'm hoping you'll get another pull request "soon" for some changes we've made to allow cashing of compiled specs in ETS). I know that technically it would be possible to do this with a single callback function but I just think this is cleaner than ```Callback(init, ...)/Callback(request, ...)```

**Comment by andreineculau** [comment]
Foreword: after my comments yesterday, I was actually thinking that for a minimal `request` callback, we could even stick to httpc (and ditch hackney, lhttpc, etc). Whoever wants "better", plug in smth else. Not sure why @isakb went with lhttpc in the beginnings. Thoughts? cc @sstrigler

> I chose to go with a module based callback because the client may need some initialization to take place and I'd rather not do that at every call.

And that initialization is not possible before calling `katt:run` ?

> I'm hoping you'll get another pull request "soon" for some changes we've made to allow cashing of compiled specs in ETS

Bring them up. I also have a small change (not published yet) to allow passing a `#katt_blueprint` instead of a filename, but I'll hold up with that until you file the PR.

**Comment by tjarvstrand** [comment]
> And that initialization is not possible before calling ```katt:run``` ?
Sure it is. I think the current implementation presents a  cleaner interface but I don't have strong opinions and I could change it if you want.

> Bring them up. I also have a small change (not published yet) to allow passing a #katt_blueprint instead of a filename, but I'll hold up with that until you file the PR.

It's actually not my changes. I believe @dszoboszlay will submit when he gets the chance to apply his changes to the latest version of KATT.

**Comment by andreineculau** [comment]
`bin/katt` crashes in this PR. Added a simple test_cli target to make sure bin/katt is not broken in 7a9fabb (now on master).

**Comment by andreineculau** [comment]
Correct me if I'm wrong, but I think #38 should cover your needs and more:

* make with ./rebar/BARE_MODE
* add jsx (or any other json parser lib), maybe tdiff too, to your codebase
* pass on a request callback that uses lhttpc, and maybe an optional parse callback (if you don't have jsx and/or tdiff)

What I hold against this PR:

* bin/katt crashes - the proper fix would be not to build it at all
* introduces a secondary callback pattern (a module option)
* unclear why katt should care about initializing the http_client (rather dealing with prerequisites before `katt:run` altogether
* "quite bloated dependency tree and we're trying to not make it explode too much for many reasons" should apply to all non-mandatory deps, including jsx tdiff, not just hackney

**Comment by tjarvstrand** [comment]
> Correct me if I'm wrong, but I think #38 should cover your needs and more

Not really since Hackney is still called explicitly in the code.

> bin/katt crashes
I can fix that.

> introduces a secondary callback pattern (a module option)

Don't really see why you have a problem with this. Like I said, I don't have any strong opinions on changing it to a function.

> Unclear why katt should care about initializing the http_client (rather dealing with prerequisites before katt:run altogether

Sure, I only added that since it was already there for Hackney and I didn't want to break backward compatibility. I don't see any need keeping it around for any other reason.

> "quite bloated dependency tree and we're trying to not make it explode too much for many reasons" should apply to all non-mandatory deps, including jsx tdiff, not just hackney

Again, I don't really see your problem with this. What's wrong with fixing one issue at a time? I

**Comment by tjarvstrand** [comment]
To clarify, sending in a custom request function would work I guess, but I would rather avoid re-implementing the entire function when the only thing we want to change is the http client request call.

**Comment by isakb** [comment]
lhttpc was the default choice really, already used in the codebase. Didn't
know about hackney's existence.

**Comment by andreineculau** [comment]
> Not really, since Hackney is still called explicitly in the code.

Update #38 so that xref passes i.e. no more hackney calls, if that's what you meant. On that line, if you rebase on top of df6dfbc, add the following to your `.travis.yml`

```yml
  - make clean .rebar/NO_HTTP_CLIENT # TEST NO_HTTP_CLIENT AS WELL
  - make
  - make .rebar/DEV_MODE deps xref
```

> What's wrong with fixing one issue at a time?

So, based on the solution of this PR and its arguments, 2 more PRs may come this way sooner or later: one for NO_JSON_PARSER, and one for NO_TEXT_DIFF, all 3 sharing the same arguments, yet we shouldn't fix the core issue i.e. no built-in extensions ?

From my own POV, it's black and white, either you want the "default" functionality with all the bells and whistles, or you decide it's too much and you stick to a core skeleton and you customize/extend at your own will.

> I would rather avoid re-implementing the entire function when the only thing we want to change is the http client request call.

Question is: do you use `x-katt-request-timeout` or `x-katt-request-sleep` in your APIBs? Because if you don't, then there's nothing to re-implement.

**Comment by tjarvstrand** [comment]
> Question is: do you use x-katt-request-timeout or x-katt-request-sleep in your APIBs? Because if you don't, then there's nothing to re-implement.

Turns out we don't so I guess we can just use the request fun off the bat. Thanks

**Comment by andreineculau** [comment]
Close or continue the discussion?

**Pull request closed without merging**